bson/dbref.py

```diff
@@ -85,7 +86,15 @@ def __getattr__(self, key):
     # infinite recursion since we override
     # __getattr__.
     def __setstate__(self, state):
-        self.__dict__.update(state)
+        for slot, value in state.items():
+            setattr(self, slot, value)
+
+    def __getstate__(self):
+        def mangle_name(n):
+            return "_DBRef"+n
+        return {mangle_name(s): getattr(self, mangle_name(s)) for s in
+                DBRef.__slots__ if
+                hasattr(self, mangle_name(s))}
```

Comment:
Would it be possible to reduce the code duplication between all these classes by defining a generic getstate/setstate function in once place?:
```python
def getstate_slots(self):
    ...
def setstate_slots(self, state):
    ...
```

Then we can reuse them like this:
```python
class DBRef(object):
    ...
    __getstate__ = getstate_slots
    __setstate__ = setstate_slots
```

Reply:
That makes sense to me. In what location should I put it?

Reply:
Done.